### DISCUSSION TITLE
全局变量异常

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#4319] [Q&A]
### Xmake 版本

xmake v2.8.3+dev.654050350

### 操作系统版本和架构

win10

### 描述问题

例如：
a = "qqq"
local b = "www"

target("xxx")
    after_build(function(target)
        print(a)--nil
        print(b)--"www"
    end)
target_end()
另外也没有_G全局表

### 期待的结果

全局变量可以所有xmake和阶段访问到

### 工程配置

无

### 附加信息和错误日志

无

**Comment by waruqi** [comment]
不要跨脚本域访问描述域变量，我限制了。。

不要访问 _G ，我也限制了。。

**Reply by PeakRacing** [reply]
局部变量不会进行限制吧？例如上述的 local b = "www"，脚本域可以正常访问，很多功能目前基于这个做的

**Reply by waruqi** [reply]
不会，你只要不是脚本域里面访问描述域的东西 就没事。。

**Reply by PeakRacing** [reply]
那应该算有问题，现在是有些target需要编译时候指定编译什么taget，target目录在哪，xmake.lua在哪引用等等，走不了onconfig，顺序太靠后了，不得不走环境变量，描述域取环境变量设置这些，脚本域也读这些变量

**Reply by waruqi** [reply]
环境变量，脚本域也是可以读的，又不是非得 描述域。。

另外，描述域的 option/set_config/set_value 等接口配置，在脚本域也是可以通过 get_config/target:values() 去读取的，我只是限制了 lua 原生变量的跨描述域访问。

**Reply by PeakRacing** [reply]
环境变量脚本域是可以读，但是很多地方要用，都写一遍处理这样代码比较臃肿，不过这不是主要问题
 option/set_config/set_value 接口生效慢，编译时候指定编译什么taget，target目录在哪，xmake.lua在哪引用这种上来描述域就要，有没有什么不走环境变量的方式？

**Reply by waruqi** [reply]
怎么个生效慢法？只有这些接口。。option/on_config 都是一次配置，到处可用，set_value 跟直接 lua 变量设置没什么区别，哪里慢了？

```lua
set_values("var", 1)

target("test")
    on_load(function (target)
          print(target:values("var"))
    end)
```

跟你直接全局 lua 变量，跨描述域访问有啥区别呢？

> 环境变量脚本域是可以读，但是很多地方要用，都写一遍处理这样代码比较臃肿

你想要全局变量一次设置，也可以：

```lua
set_values("var", os.getenv("VAR"))

target("test")
    on_load(function (target)
          print(target:values("var"))
    end)
```

另外，你走 import 模块，把常用操作，配置获取，都封装到独立模块中维护，也不用每个地方都去写一遍。。

```lua
set_moduledirs("xmake")
target("test")
    on_load(function (target)
          import("myconfig")
          print(myconfig.get("xxxx"))
    end)
```

**Reply by PeakRacing** [reply]
不是的，我没表达清楚，我现在大概是这样写的

` 
name = os.getenv("NAME"))
include(name)
target(name)
    ......
`
通过给定的NAME环境变量动态的决定加载什么目录下的什么项目，target什么名等
我之前想统一改成config方式，这样还能用menuconfig，但是config阶段太慢了，load
阶段之后才执行config阶段，所以只能还是使用环境变量上述方式，所以问有没有不使用环境变量方式实现

**Reply by waruqi** [reply]
不推荐环境变量，推荐还是走 option 处理，load / config 阶段又不会慢到哪去，又不是超大工程，load 上千个 target ，也不用 1s  里面就执行 config，我不知道你说的慢是啥。。

target 名，目录都可以在脚本域处理。。目前设计的标准做法就是走 option/config

**Reply by PeakRacing** [reply]
我说的慢指的是阶段靠后，不是速度慢，我之前问过这个问题，你说的是taget和dep这些必须编译时候确定下来，load阶段都不行，更别说config阶段，那就太晚了
![image](https://github.com/xmake-io/xmake/assets/24759100/5fbdc860-5815-4c67-a869-964bf754827f)
![image](https://github.com/xmake-io/xmake/assets/24759100/63669b93-9aa2-45bd-a85e-3f953481ac9e)
项目中这些 includes(project_dir)目录是要动态获取的，应为客户要编译sdk不同demo不同项目，包括每个target的名字，和dep依赖的target

仓库
https://gitee.com/openLuat/luatos-soc-2023

**Reply by PeakRacing** [reply]
另外在描述域中set_values设置有没有对应的获取呢？

**Reply by waruqi** [reply]
脚本域走 target:values 获取，描述域获取，就没必要走 set_values ，直接 lua 变量不就好了

**Reply by waruqi** [reply]
阶段滞后，得看你这的需求来定了，

> 项目中这些 includes(project_dir)目录是要动态获取的，应为客户要编译sdk不同demo不同项目，包括每个target的名字，和dep依赖的target

通常不建议动态 includes，而应该是全部 includes 后，根据 option/config 去动态的在 on_load 中，对 target 启用或禁用

> 包括每个target的名字，和dep依赖的target

dep 是可以在 on_load 里面动态设置的。。只是不能在 on_config 中设置而已。

**Reply by PeakRacing** [reply]
> 阶段滞后，得看你这的需求来定了，
> 
> > 项目中这些 includes(project_dir)目录是要动态获取的，应为客户要编译sdk不同demo不同项目，包括每个target的名字，和dep依赖的target
> 
> 通常不建议动态 includes，而应该是全部 includes 后，根据 option/config 去动态的在 on_load 中，对 target 启用或禁用
> 
> > 包括每个target的名字，和dep依赖的target
> 
> dep 是可以在 on_load 里面动态设置的。。只是不能在 on_config 中设置而已。

你说的我知道是什么意思，但是这是写xmake时候就已经知道要编译什么并且熟悉xmake的情况下，我们做的csdk是给所有人用的，这些人只想写自己的xmake.lua，并不会深入了解，他们习惯了make,cmake,scons这种只需要修改他们自己的编译文件的方式，而且这种也确实是很方便维护和上手的，所以只维护自己的xmake.lua我觉得应该是合理的情况，其他的sdk基本也都是这么做的，包括idf,rtthread等等，他们用的make,cmake,scons等，放到xmake中应该就是include使用者自己的xmake.lua，如果这个阶段一定要最前面，config在后面，那还是要自己走环境变量实现，但是这就脱离了xmake，我认为一个完善的构建应该脱离环境变量，所以我一直在找各种方法，把之前的环境变量都换成了option，但是option+config阶段靠后这里是在绕不过去……

**Reply by waruqi** [reply]
你 csdk 对用户导出，不应该做成 package 不更好么。。用户端只需要 `add_requires("csdk", {configs = {example = get_config("example")}})` 去配置集成，都是支持 option 的

其他不用管，另外 package 里面也可以导出 custom rule...


另外， 也可以参考 https://github.com/microsoft/cheriot-rtos 的做法，只需要用户端统一一个 `includes("sdk")`

另外，我看你的工程里面这种，通过配置去选择编译 target。

```lua
target("csdk")
    set_kind("static")
    set_targetdir("$(buildir)/csdk")
    add_deps(USER_PROJECT_NAME)

target(USER_PROJECT_NAME..".elf")
    set_kind("binary")
    set_targetdir("$(buildir)/"..USER_PROJECT_NAME)
    add_deps("csdk")
    add_deps(USER_PROJECT_NAME)
```

完全可以一次性 includes 后，根据 set_default + option 去启用，不走环境变量。。

先 includes 所有

```lua
includes("project")
```

project/xmake.lua 里面，遍历所有 example ，一次性定义完所有 target，并默认 default = false 禁用

```lua
includes("*/xmake.lua")

for _, filepath in ipairs(os.files(path.join(os.scriptdir(), "*", "xmake.lua"))) do
    local name = path.basename(path.directory(filepath))
    target("csdk_" .. name)
       set_kind("static")
       set_default(false)
       set_targetdir("$(buildir)/csdk")
       add_deps(name)

    target(name..".elf")
      set_kind("binary")
      set_targetdir("$(buildir)/".. name)
      add_deps("csdk_" .. name)
      add_deps(name)
      set_default(false)
      on_config(function (target)
           -- 如果 xmake f --example=xxx 用户配置启用当前 target，就单独编译它，其他的默认禁用
           if is_config("example", name) then
               target:set("default", true)
           end
      end)
      -- 其他所有 通用配置，可以通过 rules 统一配置
      add_rules("common_configuration") 
      -- 也可以在描述域，通过 function 去 wrap 通用的 add_defines,add_links, on_load 处理，统一设置，怎么搞都行。。
      add_common_configuration()
end
```

反正，xmake 的设计就是，includes 只管 includes 所有，仅仅只是整合所有的子工程配置，不做选择编译和状态控制，target 的编译选择，都是 set_default 和 option 干的事情，跟 includes 没有任何关系。。

所有复杂的控制处理都是 脚本域+rule 里面去处理的

**Reply by PeakRacing** [reply]
反向include,这是个好办法，但是现在这个仓库改动的话变动太大，客户肯定了受不了，先这样吧，这个只能先走环境变量了，下一个sdk再改，感谢回复

**Comment by PeakRacing** [comment]
> 不要跨脚本域访问描述域变量，我限制了。。
> 
> 不要访问 _G ，我也限制了。。
